### DISCUSSION TITLE
Mikro-orm v6 serialization with em.populate()

### DISCUSSION THREAD

**Discussion by Zmaon** [mikro-orm/mikro-orm#5386] [General]
Hi,

We have a case where we need to use `em.populate()` to load a lot of field on existing entities in some conditions. Following this change (https://mikro-orm.io/docs/upgrading-v5-to-v6#implicit-serialization-changes), it seems that those fields are not serialized by default, and that we should manually do a `wrap().populated()` for each of them in order to have them in the output (it's a Nest.js project). 

Shouldn't a call to `em.populate()` automatically add the fields in the serialization? It seems to be a lot of boilerplate to do it manually, when the intent seems quite clear when you populate this way. 

Thanks!

Simon

**Comment by B4nan** [comment]
> Shouldn't a call to em.populate() automatically add the fields in the serialization?

It's all driven by the populate hint of the root entity you serialize, if you call `em.populate` on it, it should work out of box I think. Having that on a child is not enough, the whole point of this is that there need to be one root that knows how the entity graph should look like, otherwise there is no way to properly serialize cycles.

Can you provide some repro so its clear what you are doing?

edit: this is the method responsible for updating the serialization context:

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/entity/EntityLoader.ts#L122C11-L122C34

**Comment by Zmaon** [comment]
I see. Digging further I think that our issue is only related to serialization, not population. 

To go into more detail our problem is that we populate a tree where we can have an arbitrary number of branches. Unfortunately I can't share the code, but if I simplify we have an entity like that:

```JS
class Entity {
  @OneToMany(() => Entity, entity => entity.parent)
  children = new Collection<Entity>this);

  @ManyToOne(() => Entity, { nullable: true })
  parent? = Entity;

  @Property({ type: 'varchar', length: 100 })
  name!: string;

  # A bunch of other properties
  ...
}
```

Before version 6, loading all the entities with a `findAll()` while populating only the first level was directly returning the whole tree:

```JS
 const entities: Entity[] = await this.entitityRepository.findAll({
      populate: [
        'parent',
        'children',
        ...
      ]
  })

# the interface returns the whole tree
return entities
```

With version 6, it only returns the first level, because of the change on the serialize(). Is there any way to tell the serializer to take everything into account as it was before?

**Reply by B4nan** [reply]
Before we dive deeper here, have you considered using explicit serialization instead? As that is often the best thing to do, you know well how the DTO should be shaped, right?

**Reply by Zmaon** [reply]
I did try and ended up with something like that:
```JS
  return serialize(entities, {
    populate: ['children.children.children.children'],
  }) as unknown as Entity[];
```
So I need to know how many levels there are, which is not good. Do you see a better way?

**Reply by B4nan** [reply]
Hmm, so you don't know how the shape will look like? 

> Do you see a better way?

Maybe try using `wrap(e).toPOJO()` which ignores the serialization hints completely.

**Reply by Zmaon** [reply]
That's the thing, we don't know the shape of the output, which happens with this kind of self-referencing entities where you can have any number of nested levels. 

I did a quick tryout with `wrap(e).toPOJO()`, which seems to accept only one entity at a time, no luck so far.

By the way may I ask what was the motivation for serializing only what is explicitly defined in the populate? Does it also mean that relations defined as `eager: true` are not serialized? I guess there are good reasons for this, but it seems to loose some flexibility.

**Reply by B4nan** [reply]
> I did a quick tryout with wrap(e).toPOJO(), which seems to accept only one entity at a time, no luck so far.

How is that a problem? You have an array, just map it to the `toPOJO` calls.

> By the way may I ask what was the motivation for serializing only what is explicitly defined in the populate?

To have a predictable shape. Without it, you can end up with some already loaded entity resulting in object when it wasn't populated, breaking a stable response.

> Does it also mean that relations defined as eager: true are not serialized? I guess there are good reasons for this, but it seems to loose some flexibility.

Eager loading is implemented by automatic computation of the populate hint, it's the same as if you would put the relation there explicitly. So nope, that is not a problem.

**Reply by Zmaon** [reply]
> How is that a problem? You have an array, just map it to the toPOJO calls.

Actually it crashes. I need more time to check why. 

> To have a predictable shape. Without it, you can end up with some already loaded entity resulting in object when it wasn't populated, breaking a stable response.

I understand the motivation to have a more predictable output, we actually used class transformers in Nest.js so far to achieve that. Still in our use case it would be really nice to have a way to explicitly ask for "any shape". 

> Eager loading is implemented by automatic computation of the populate hint, it's the same as if you would put the relation there explicitly. So nope, that is not a problem.

Ok nice.

**Reply by B4nan** [reply]
> Still in our use case it would be really nice to have a way to explicitly ask for "any shape".

That's exactly the `toPOJO()` call, unless you reach a cycle, it will follow all populated objects. I guess the cycle detection fails if it crashes.

**Reply by B4nan** [reply]
Actually, there is also `serialize(..., { populate: ['*'] })` which has a bit different implementation than the `toPOJO` helper, so try that as well.

**Reply by Zmaon** [reply]
Interestingly `serialize(..., { populate: ['*'] })` does work, but the API returns an insane amount of data (~280MB) when using it which renders it unusable. I guess that this is because of the circular relation parent<->children.

**Reply by B4nan** [reply]
Well, that's why it doesn't work like that by default... :] 

I think the idea of returning a random payload shape is flawed, you should rethink this approach.

**Reply by Zmaon** [reply]
Yes, I'm now wondering how it was working so well with the previous version of Mikro-orm.

**Reply by Zmaon** [reply]
> I think the idea of returning a random payload shape is flawed, you should rethink this approach.

Any tree would have this problem if the number of levels is unknown, and trees are pretty common in API answers. In my opinion an ORM shouldn't be too opinionated and it should leave its users the freedom to shape their payloads in the way they want. 

This said I understand that the problem is not trivial, and I also think that a predictable output is a good thing. I just wish I had a simple solution for my problem.

**Reply by B4nan** [reply]
> In my opinion an ORM shouldn't be too opinionated and it should leave its users the freedom to shape their payloads in the way they want.

You can shape your payload in any way you want, you can serialize stuff yourself instead of depending on the implicit serialization. This is not about being opinionated, this is about being predictable. You want something else, so you need to provide custom code to do that. The previous implementation wasn't giving users freedom, in my eyes it was the opposite. Check [the PR](https://github.com/mikro-orm/mikro-orm/pull/4203) and the issues it closed to have a better understanding of what changed and why.

**Reply by Zmaon** [reply]
Ok we ended up implementing a custom serialization for the tree. 

Thanks a lot for your prompt answers, really appreciated.

**Reply by B4nan** [reply]
If you can think of something we could add to the ORM toolbelt, I am open for suggestions. Maybe some option (for explicit serialization) to override the function that says if something should be populated or not would do the job? 

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/serialization/EntitySerializer.ts#L26
